### PR TITLE
Add blank line check between package and import statements

### LINKED ISSUE #3251: Enforce blank line between `package` and `import` statements
## Expected Rule behavior

When linting the following, I would expect a style violation due the lack of a blank line between the `package` and `import` statements:

```kotlin
package com.example
import foo.bar
```

## Additional information
* Current version of ktlint: 1.8.0
* Styleguide section: https://developer.android.com/kotlin/style-guide#structure

### PR DESCRIPTION
**Pull request by maheen8q**
## Description

Add blank line check between package and import statements

The Android Kotlin Style Guide requires exactly one blank line between 
the package statement and import statements:
https://developer.android.com/kotlin/style-guide#structure

ktlint was not enforcing this rule. This PR extends the existing 
`PackageNameRule` to also check for the missing blank line, with 
autocorrect support.

Fixes #3251 

## Checklist

Before submitting the PR, please check following (checks which are not relevant may be ignored):
- [✅] Commit message are well written. In addition to a short title, the commit message also explain why a change is made.
- [✅] At least one commit message contains a reference `Closes #<xxx>` or `Fixes #<xxx>` (replace`<xxx>` with issue number)
- [✅] Tests are added
- [ ] KtLint format has been applied on source code itself and violations are fixed
- [✅] PR title is short and clear (it is used as description in the release changelog)
- [✅] PR description added (background information)

[Documentation](https://pinterest.github.io/ktlint/) is updated. See [difference between snapshot and release documentation](https://github.com/pinterest/ktlint/tree/master/documentation)
- [] [Snapshot documentation](https://github.com/pinterest/ktlint/tree/master/documentation/snapshot) in case documentation is to be released together with a code change
- [ ] [Release documentation](https://github.com/pinterest/ktlint/tree/master/documentation/release-latest) in case documentation is related to a released version of ktlint and has to be published as soon as the change is merged to master

### CODE REVIEW

File: ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageNameRuleTest.kt [hunk 1]
```diff
@@ -88,4 +88,47 @@ class PackageNameRuleTest {
...
+        packageNameRuleAssertThat(code)
+            .hasLintViolation(
+                line = 2,
+                col = 1,
+                detail = "Missing blank line between package statement and import statements",
+            )
+    }
+
+    @Test
+    fun `autocorrects missing blank line between package and import`() {
```
Comment (paul-dingemans) [inline ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageNameRuleTest.kt]: This test should be merged with previous test. There is no need to have separate tests for report and fixing.

**Review by paul-dingemans — CHANGES_REQUESTED** [review]
Tnx for giving this a try so quickly. Please try to adhere more with existing code patterns and standards in current ktlint code.

**Comment by maheen8q** [comment]
Review comments addressed, please take another look

File: ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageNameRule.kt [hunk 2]
```diff
@@ -37,11 +40,34 @@ public class PackageNameRule : StandardRule("package-name") {
...
+
+        // https://developer.android.com/kotlin/style-guide#structure
+        node
+            .takeIf { it.elementType == PACKAGE_DIRECTIVE }
+            ?.let { packageNode ->
+                val nextSibling = packageNode.treeNext ?: return
+                val nextCodeSibling = packageNode.nextCodeSibling20
+                if (nextCodeSibling?.elementType == IMPORT_DIRECTIVE) {
+                    val newlineCount = nextSibling.text.count { it == '\n' }
+                    if (newlineCount < 2) {
```
Comment (paul-dingemans) [inline ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageNameRule.kt]: The style guide describes that there should be exactly 1 blank line between package and imports. So test on on "newLineCount != 2.

Also, tests corner case with comment above imports
```
package foo

// Some comment
import bar.*
```

**Review by paul-dingemans — CHANGES_REQUESTED** [review]
Tnx for follow-up. I have some more remarks.

**Comment by maheen8q** [comment]
Hi @paul-dingemans , thanks for the feedback! I've moved the blank line check into a separate PackageImportSpacingRule as suggested. Here's a summary of all changes made:

Moved blank line between package and import check to new PackageImportSpacingRule
Cleaned up PackageNameRule to only handle package naming validation
Used newlineCount != 2 to enforce exactly 1 blank line
Created PackageImportSpacingRuleTest with all test cases including:

No blank line between package and import
Too many blank lines between package and import
Comment above import edge case


Removed blank line tests from PackageNameRuleTest

Please take another look, happy to make any further changes!

File: ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageImportSpacingRule.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                            nextSibling.startOffset,
+                            "Missing blank line between package statement and import statements",
+                            true,
+                        ).ifAutocorrectAllowed {
+                            val corrected = "\n\n" + nextSibling.text.trimStart('\n')
+                            (nextSibling as? LeafElement)?.rawReplaceWithText(corrected)
+                        }
+                    }
+                }
+            }
```
Comment (paul-dingemans) [inline ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/PackageImportSpacingRule.kt]: Yesterday, I did write remark below, but unfortunately forgot to add it to the review.

Probably this can be simplified, and be made more robust. Please, have a look at the AST below:
Image

The `PACKAGE_DIRECTIVE` and `IMPORT_LIST` nodes always exist, but they can be empty. A comment can be added between package and first import statement. A better approach would be following:

```
        node
            .takeIf { it.elementType == PACKAGE_DIRECTIVE }
            ?.siblings()
            ?.takeWhile { it.elementType != IMPORT_LIST }
            ?.firstOrNull { it.isWhiteSpaceWithNewline }
            ?.takeIf { whitespace -> whitespace.text.count { it == '\n' } != 2 }
            ?.let { whitespace ->
                emit(
                    whitespace.startOffset,
                    "Expected exactly one blank line between package statement and import statements",
                    true,
                ).ifAutocorrectAllowed { whitespace.replaceTextWith("\n\n") }
            }
```

**Review by paul-dingemans — CHANGES_REQUESTED** [review]

**Comment by maheen8q** [comment]
Hi Paul, thank you for the detailed feedback and the AST explanation!
I have updated the PR with the following changes:
PackageImportSpacingRule:
Implemented the RuleV2.Experimental marker interface as requested.
Updated the @SinceKtlint annotation to ("2.0", EXPERIMENTAL).
Refactored the logic to use the siblings() and takeWhile { it.elementType != IMPORT_LIST } approach. This is much cleaner and handles intermediate comments correctly.
PackageNameRule:
Reverted all changes to this rule, including the regex and the it.elementType refactor, to keep this PR focused and consistent with the existing codebase.
Let me know if there is anything else you'd like me to adjust.

**Comment by paul-dingemans** [comment]
Hmm, I had to make quite some changes to make it actually work. Did you run the tests on your local machine? Also, documentation was missing.
